### PR TITLE
Adds autocompletion for afan

### PR DESCRIPTION
**Pull request by chinmaydd**
We had tab based autocompletion for "afvn" but not for "afan".

### CODE REVIEW

File: libr/core/core.c [hunk 1]
```diff
@@ -604,6 +604,28 @@ static int autocomplete(RLine *line) {
...
+			f_ptr = r_sub_str_lchr (line->buffer.data, 0, line->buffer.index, ' ');
+			f_ptr = f_ptr != NULL ? f_ptr + 1 : line->buffer.data;
+			l_ptr = r_sub_str_rchr (line->buffer.data, line->buffer.index, len, ' ');
+			if (l_ptr == NULL) {
+				l_ptr = line->buffer.data + strlen (line->buffer.data);
+			}
+
+			r_list_foreach (vars, iter, var) {
+				if (!strncmp (f_ptr, var->name, l_ptr - f_ptr)) {
+					tmp_argv[j++] = strdup(var->name);
```
Comment (radare) [inline libr/core/core.c]: Missing space before (

File: libr/core/core.c [hunk 2]
```diff
@@ -604,6 +604,28 @@ static int autocomplete(RLine *line) {
...
 			line->completion.argv = tmp_argv;
+		} else if ((!strncmp (line->buffer.data, "afan ", 5))) {
+			RAnalFunction *fcn = r_anal_get_fcn_in (core->anal, core->offset, 0);
+			RList *vars = r_anal_var_list (core->anal, fcn, R_ANAL_VAR_KIND_ARG);
+			const char *f_ptr, *l_ptr;
+			RAnalVar *var;
+			int j = 0, len = strlen (line->buffer.data);
+
+			f_ptr = r_sub_str_lchr (line->buffer.data, 0, line->buffer.index, ' ');
+			f_ptr = f_ptr != NULL ? f_ptr + 1 : line->buffer.data;
```
Comment (radare) [inline libr/core/core.c]: Unnecessary !=NULL

File: libr/core/core.c [hunk 3]
```diff
@@ -604,6 +604,28 @@ static int autocomplete(RLine *line) {
...
+			l_ptr = r_sub_str_rchr (line->buffer.data, line->buffer.index, len, ' ');
+			if (l_ptr == NULL) {
```
Comment (radare) [inline libr/core/core.c]: Use !l_ptr

File: libr/core/core.c [hunk 4]
```diff
@@ -604,6 +604,28 @@ static int autocomplete(RLine *line) {
...
+				l_ptr = line->buffer.data + strlen (line->buffer.data);
```
Comment (radare) [inline libr/core/core.c]: Iirc u dont need to strlen this bugfer because theres a var that caches the length

File: libr/core/core.c [hunk 5]
```diff
@@ -604,6 +604,28 @@ static int autocomplete(RLine *line) {
 			tmp_argv[j] = NULL;
 			line->completion.argc = j;
 			line->completion.argv = tmp_argv;
+		} else if ((!strncmp (line->buffer.data, "afan ", 5))) {
+			RAnalFunction *fcn = r_anal_get_fcn_in (core->anal, core->offset, 0);
+			RList *vars = r_anal_var_list (core->anal, fcn, R_ANAL_VAR_KIND_ARG);
+			const char *f_ptr, *l_ptr;
+			RAnalVar *var;
+			int j = 0, len = strlen (line->buffer.data);
```
Comment (radare) [inline libr/core/core.c]: Unnecessary strlen iirc

**Comment by ret2libc** [comment]
Hi! As you said there is already autocompletion for `afvn`, so it would be better to avoid the duplication of code and just handle both cases here https://github.com/radare/radare2/blob/master/libr/core/core.c#L585 . You can handle `afvn` and `afan` in that case and pass R_ANAL_VAR_KIND_VAR or R_ANAL_VAR_KIND_ARG to `r_anal_var_list` depending on the command. The rest of the code should work for both types.

**Comment by radare** [comment]
agree with @ret2libc its better to reuse

**Comment by Maijin** [comment]
@oddcoder see this pls

**Comment by radare** [comment]
ping

**Comment by chinmaydd** [comment]
@radare I will complete this asap. I agree with ret2libc.

**Comment by radare** [comment]
👍

**Comment by chinmaydd** [comment]
Updated. Also, I could find out the variable you were mentioning which caches the `line->buffer.data` length. Maybe if you can point me to the right resource I can modify it.

**Comment by ret2libc** [comment]
LGTM

**Pull request merged**